File: src/skillhub/ir.py
"""SkillIR — intermediate representation for canonical SKILL.md.

设计要点:
- body: bytes (不是 str) — 零损耗硬约束。
  不做编码归一(防 \\r\\n -> \\n / BOM strip 等)。
  parser 切分 frontmatter 与 body 时, body 收 frontmatter 边界之后的所有原字节。
- frontmatter 字段: Level enum + 7 个稳定可选字段(canonical 标准 frontmatter 约束)。
- agent_overrides: 每个 Agent 专有字段副本(反向导入 Agent skill 时抽取)。
  canonical 不承载, emitter 用时按需取(从 skills/<name>/.agent_overrides/<agent>.toml)。
- requires: 能力标签列表, 跨切关注, 用于 prompt_inject。

往返零损耗的含义(roundtrip):
    parse(canonical) -> IR0
    IR0 -> emit_canonical -> bytes
    parse -> IR1
    断言 IR1.body == IR0.body (字节完全等值)
frontmatter 字段往返允许 dict 等值即可(序化格式可漂移, body 不漂移)。
"""

from __future__ import annotations

import enum
import hashlib
from dataclasses import dataclass, field
from pathlib import Path
from typing import Optional


class Level(str, enum.Enum):
    """skill 分级(决策 4.4)。

    AUTO         稳定高频, 同步 + 允许模型自动触发
    MANUAL       稳定低频, 同步 + 禁止自动触发(转 disable-model-invocation / enabled_at:false)
    EXPERIMENTAL 实验中, 同步 + 禁止自动触发(同 manual)
    DISABLE      暂停, 不同步 + 清理已部署副本(canonical 保留以便恢复)
    """

    AUTO = "auto"
    MANUAL = "manual"
    EXPERIMENTAL = "experimental"
    DISABLE = "disable"

    def __str__(self) -> str:
        return self.value


@dataclass
class SkillIR:
    """Canonical SKILL.md 的语义化中间表示。

    body 是零损耗核心: parser/emitter 全程以 bytes 流转, 不 decode/encode。
    """

    name: str
    description: str
    body: bytes                                # 零损耗核心: bytes 不 str
    level: Level = Level.AUTO
    native_agent: Optional[str] = None         # 原生 Agent(决策: emitter 注入"原生于 X"前言)
    requires: list[str] = field(default_factory=list)   # 能力标签, 对 capabilities.toml
    description_zh: Optional[str] = None      # 双语 desc(canonical 共额外, emitter 镜像到 _cn/_zh)
    name_zh: Optional[str] = None             # 双语 name
    version: Optional[str] = None
    license: Optional[str] = None
    agent_overrides: dict[str, dict] = field(default_factory=dict)
    resources: list[Path] = field(default_factory=list)
    source_path: Optional[Path] = None         # parser 溯源(写 manifest 用)

    def body_hash(self) -> str:
        """body 的 sha256 — manifest ir_hash 字段用, 跨机器确认零损耗。"""
        return "sha256:" + hashlib.sha256(self.body).hexdigest()

    def level_allows_sync(self) -> bool:
        """disable 不参与同步, 只保留 canonical。"""
        return self.level != Level.DISABLE

    def level_allows_auto_trigger(self) -> bool:
        """manual/experimental/disable 都禁止模型自动触发(转 Agent frontmatter 字段)。"""
        return self.level == Level.AUTO

    def to_frontmatter_dict(self) -> dict:
        """canonical frontmatter 字段 dict(供 emitter 序化回 SKILL.md)。"""
        fm: dict = {"name": self.name, "description": self.description, "level": self.level.value}
        if self.native_agent is not None:
            fm["native_agent"] = self.native_agent
        if self.requires:
            fm["requires"] = list(self.requires)
        if self.description_zh is not None:
            fm["description_zh"] = self.description_zh
        if self.name_zh is not None:
            fm["name_zh"] = self.name_zh
        if self.version is not None:
            fm["version"] = self.version
        if self.license is not None:
            fm["license"] = self.license
        return fm

    @classmethod
    def from_frontmatter_dict(cls, fm: dict, body: bytes, source_path: Optional[Path] = None) -> "SkillIR":
        """从 canonical frontmatter dict + body bytes 构建 IR。

        允许 frontmatter 含未识别字段(留作 leftover, 不抛错)。
        """
        # required
        name = fm["name"]
        description = fm["description"]
        level = Level(fm.get("level", "auto"))
        return cls(
            name=name,
            description=description,
            body=body,
            level=level,
            native_agent=fm.get("native_agent"),
            requires=list(fm.get("requires") or []),
            description_zh=fm.get("description_zh"),
            name_zh=fm.get("name_zh"),
            version=fm.get("version"),
            license=fm.get("license"),
            source_path=source_path,
        )

File: internal/ir/ir.go
// Package ir — canonical SKILL.md 的语义化中间表示。
//
// 移植合同(对应 Python ir.py, 语义逐条等价):
//   - Body 是 []byte 零损耗硬约束: parser/emitter 全程字节流转, 不 decode 不 normalize
//   - FMRaw/FMOrig 保 frontmatter 原始字节与原始 dict, emitter 据此做字段级透传
//     (未变更字段原字节保留, 引号/顺序不漂)
//
// Roundtrip 零损耗含义: parse → IR0 → emit → parse → IR1, IR1.Body == IR0.Body 字节等值;
// frontmatter 往返允许 dict 等值(序化格式可漂, body 不漂)。
package ir

import (
	"crypto/sha256"
	"encoding/hex"
)

// Level — skill 分级(决策 4.4)。
type Level string

const (
	// Auto 稳定高频, 同步 + 允许模型自动触发
	Auto Level = "auto"
	// Manual 稳定低频, 同步 + 禁止自动触发
	Manual Level = "manual"
	// Experimental 实验中, 同步 + 禁止自动触发(同 manual)
	Experimental Level = "experimental"
	// Disable 暂停, 不同步 + 清理已部署副本(canonical 保留以便恢复)
	Disable Level = "disable"
)

// ValidLevel — canonical frontmatter level 合法值。
var ValidLevels = []Level{Auto, Manual, Experimental, Disable}

func (l Level) Valid() bool {
	switch l {
	case Auto, Manual, Experimental, Disable:
		return true
	}
	return false
}

// AllowsSync — disable 不参与同步, 只保留 canonical。
func (l Level) AllowsSync() bool { return l != Disable }

// AllowsAutoTrigger — manual/experimental/disable 都禁止模型自动触发。
func (l Level) AllowsAutoTrigger() bool { return l == Auto }

// SkillIR — canonical SKILL.md 的 IR。
//
// 可选字段用 *string(非 string): Python 原版用 None 区分「缺省」与「空串」
// (`description_zh: ''` 是存在且值为空, 透传时必须保留), Go string 合并两者
// 会把显式空串误判为删除 → 字节漂移。
type SkillIR struct {
	Name        string
	Description string
	Body        []byte   // 零损耗核心: 原字节, 不 decode
	Level       Level    // 缺省 auto(FromFrontmatterDict 填)
	NativeAgent *string  // nil = 无
	Requires    []string // 能力标签, 仅文档标注; 空(含显式 requires: [])不输出, 对应 Python `if self.requires:`
	DescZH      *string  // 双语 desc(_zh; TeleAgent 镜像成 _cn)
	NameZH      *string  // 双语 name
	Version     *string
	License     *string
	Source      *string // 来源 provenance(git URL); nil = 本地导入/未知
	SourcePath  string    // parser 溯源
	// 字段级透传: parser 保留 frontmatter 原始字节 + 原始 dict。
	// FMRaw 为 nil(如 import 新建)时 emitter 走全量 dump。
	FMRaw  []byte         // frontmatter 原始文本(不含 --- 边界); nil = 无
	FMOrig map[string]any // 原始解析出的 frontmatter dict; FMRaw 非 nil 时必有
}

// BodyHash — body sha256(manifest ir_hash 用, 跨机零损耗验证)。
func (ir *SkillIR) BodyHash() string {
	sum := sha256.Sum256(ir.Body)
	return "sha256:" + hex.EncodeToString(sum[:])
}

// ToFrontmatterDict — canonical frontmatter 字段 dict(emit 回 SKILL.md 用)。
// 字段顺序即声明序(新增字段 dump 时保持此序)。
func (ir *SkillIR) ToFrontmatterDict() map[string]any {
	fm := map[string]any{
		"name":        ir.Name,
		"description": ir.Description,
		"level":       string(ir.Level),
	}
	if ir.NativeAgent != nil {
		fm["native_agent"] = *ir.NativeAgent
	}
	if len(ir.Requires) > 0 {
		fm["requires"] = append([]string{}, ir.Requires...)
	}
	if ir.DescZH != nil {
		fm["description_zh"] = *ir.DescZH
	}
	if ir.NameZH != nil {
		fm["name_zh"] = *ir.NameZH
	}
	if ir.Version != nil {
		fm["version"] = *ir.Version
	}
	if ir.License != nil {
		fm["license"] = *ir.License
	}
	if ir.Source != nil {
		fm["source"] = *ir.Source
	}
	return fm
}

// CanonicalFieldOrder — canonical 字段产出序(新增字段 dump 追加序, 对应 Python dict 声明序)。
var CanonicalFieldOrder = []string{
	"name", "description", "level", "native_agent", "requires",
	"description_zh", "name_zh", "version", "license", "source",
}

// FromFrontmatterDict — 从 frontmatter dict + body bytes 构建 IR。
// 允许未识别字段(留作 leftover, 不报错)。fmRaw 非 nil 时存入供字段级透传。
// 缺 name/description 返回错误(canonical 必填)。
func FromFrontmatterDict(fm map[string]any, body []byte, sourcePath string,
	fmRaw []byte) (*SkillIR, error) {
	name, _ := fm["name"].(string)
	desc, _ := fm["description"].(string)
	if name == "" {
		return nil, &RequiredFieldError{Field: "name"}
	}
	if desc == "" {
		return nil, &RequiredFieldError{Field: "description"}
	}

	level := Auto
	if lv, ok := fm["level"]; ok {
		if s, ok := lv.(string); ok {
			level = Level(s)
		}
	}

	ir := &SkillIR{
		Name:        name,
		Description: desc,
		Body:        body,
		Level:       level,
		SourcePath:  sourcePath,
	}
	if v, ok := fm["native_agent"]; ok {
		ir.NativeAgent = strPtr(v)
	}
	if req, ok := fm["requires"]; ok {
		if list, ok := req.([]any); ok {
			for _, e := range list {
				if s, ok := e.(string); ok {
					ir.Requires = append(ir.Requires, s)
				}
			}
		}
	}
	if v, ok := fm["description_zh"]; ok {
		ir.DescZH = strPtr(v)
	}
	if v, ok := fm["name_zh"]; ok {
		ir.NameZH = strPtr(v)
	}
	if v, ok := fm["version"]; ok {
		ir.Version = strPtr(v)
	}
	if v, ok := fm["license"]; ok {
		ir.License = strPtr(v)
	}
	if v, ok := fm["source"]; ok {
		ir.Source = strPtr(v)
	}
	if fmRaw != nil {
		ir.FMRaw = fmRaw
		fmOrig := make(map[string]any, len(fm))
		for k, v := range fm {
			fmOrig[k] = v
		}
		ir.FMOrig = fmOrig
	}
	return ir, nil
}

// RequiredFieldError — frontmatter 缺必填字段。
type RequiredFieldError struct{ Field string }

func (e *RequiredFieldError) Error() string {
	return "frontmatter 缺必填字段: " + e.Field
}

// strPtr — frontmatter 值 → *string(非 string 值视为缺省; 实际数据都是字符串)。
func strPtr(v any) *string {
	if s, ok := v.(string); ok {
		return &s
	}
	return nil
}